content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-2.md

```diff
@@ -0,0 +1,231 @@
+
+```json
+{"_id":"AA_2017-02-16_EWR_ORD_1124”,
+"dofW”:4, "carrier":"AA", "origin":"EWR",
+"dest":"ORD”, "crsdeptime":705,
+"crsarrtime”:851, "crselapsedtime":166.0,"dist":719.0}
+```
+
+![Example Use Case Data](/img/example-use-case-data.png)
+
+## Spark Kafka Consumer Producer Code
+
+**Parsing the Data Set Records**
+
+We use a Scala case class and [Structype](https://spark.apache.org/docs/latest/sql-programming-guide.html#programmatically-specifying-the-schema) to define the schema, corresponding to the input data.
+
+![Structype](/img/structype.png)
+
+**Loading the Model**
+
+The Spark [CrossValidatorModel](https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.ml.tuning.CrossValidatorModel) class is used to load the saved model fitted on the historical flight data.
+
+![Loading the Model](/img/loading-model.png)
+
+**Spark Streaming Code**
+
+These are the basic steps for the Spark Streaming Consumer Producer code:
+
+1. Configure Kafka Consumer and Producer properties.
+2. Initialize a Spark StreamingContext object. Using this context, create a DStream that reads a message from a Topic.
+3. Apply transformations (which create new DStreams).
+4. Write messages from the transformed DStream to a Topic.
+5. Start receiving data and processing. Wait for the processing to be stopped.
+
+**We will go through each of these steps with the example application code.**
+
+## **1) Configure Kafka Consumer Producer properties**
+
+The first step is to set the KafkaConsumer and KafkaProducer configuration properties, which will be used later to create a DStream for receiving/sending messages to topics. You need to set the following paramters:
+
+* Key and value deserializers: for deserializing the message.
+* Auto offset reset: to start reading from the earliest or latest message.
+* Bootstrap servers: this can be set to a dummy host:port since the broker address is not actually used by MapR Event Store.
+
+For more information on the configuration parameters, [see the MapR Event Store documentation.](https://docs.datafabric.hpe.com/62/MapR_Streams/differences_in_configuration_parameters_for_producers_and_consumers.html)
+
+![Configure Kafka](/img/configure-kafka-cosumer.png)
+
+## **2) Initialize a Spark StreamingContext object.**
+
+ConsumerStrategies.Subscribe, as shown below, is used to set the topics and Kafka configuration parameters. We use the [KafkaUtils createDirectStream](https://spark.apache.org/docs/latest/streaming-kafka-0-10-integration.html) method with a StreamingContext, the consumer and location strategies, to create an input stream from a MapR Event Store topic. This creates a DStream that represents the stream of incoming data, where each message is a key value pair. We use the DStream map transformation to create a DStream with the message values.
+
+![Initialize a Spark StreamingContext Object 1](/img/initialize-spark-streamingcontext.png)
```

Comment:
Again, check images for size and fuzziness throughout.

Reply:
I adjusted all images to 800 width